core/invoker/src/main/scala/whisk/core/containerpool/docker/DockerClientWithFileAccess.scala

```diff
@@ -0,0 +1,186 @@
+     *
+     * Does not verify that the returned file actually exists.
+     *
+     * @param containerId Id of the desired Docker container
+     * @return canonical location of the container's log file
+     */
+    protected def containerLogFile(containerId: ContainerId) =
+        new File(containerDirectory(containerId), s"${containerId.asString}-json.log").getCanonicalFile()
+
+    /**
+     * Provides the contents of the specified Docker container's configuration
+     * file as JSON object.
+     *
+     * @param configFile the container's configuration file in JSON format
+     * @return contents of configuration file as JSON object
+     */
+    protected def configFileContents(configFile: File): Future[JsObject] = Future {
+        blocking { // Needed due to synchronous file operations
+            val source = Source.fromFile(configFile)
+            val config = try source.mkString finally source.close()
+            config.parseJson.asJsObject
+        }
+    }
+
+    /**
+     * Extracts the IP of the container from the local config file of the docker daemon.
+     *
+     * A container may have more than one network. The container has an
+     * IP address in each of these networks such that the network name
+     * is needed.
+     *
+     * @param id the id of the container to get the IP address from
+     * @param network name of the network to get the IP address from
+     * @return the ip address of the container
+     */
+    protected def ipAddressFromFile(id: ContainerId, network: String): Future[ContainerIp] = {
+        configFileContents(containerConfigFile(id)).map { json =>
+            val networks = json.fields("NetworkSettings").asJsObject.fields("Networks").asJsObject
+            val specifiedNetwork = networks.fields(network).asJsObject
+            val ipAddr = specifiedNetwork.fields("IPAddress")
+            ContainerIp(ipAddr.convertTo[String])
+        }
+    }
+
+    // See implemented trait for description
+    override def inspectIPAddress(id: ContainerId, network: String)(implicit transid: TransactionId): Future[ContainerIp] = {
+        ipAddressFromFile(id, network).recoverWith {
+            case _ => super.inspectIPAddress(id, network)
+        }
+    }
+
+    // See implemented trait for description
+    def rawContainerLogs(containerId: ContainerId, fromPos: Long): Future[ByteBuffer] = Future {
+        blocking { // Needed due to synchronous file operations
+            var fis: FileInputStream = null
+            try {
+                val file = containerLogFile(containerId)
+                val size = file.length
+
+                fis = new FileInputStream(file)
+                val channel = fis.getChannel().position(fromPos)
+
+                // Buffer allocation may fail if the log file is too large to hold in memory or
+                // too few space is left on the heap, respectively.
+                var remainingBytes = (size - fromPos).toInt
```

Comment:
Created the variant below and reviewed with @cbickel and @markusthoemmes. Jointly came to the conclusion that we want to keep the current implementation to avoid the risk of changing the current behaviour by mistake. I'm going to create an issue for refactoring log file reading according to the proposal below.

```
    protected def readFile(fileChannel: FileChannel, readBuffer: ByteBuffer, remainingBytes: Int): ByteBuffer = {
        println(s"readBuffer=${readBuffer}")
        if (remainingBytes <= 0) {
            readBuffer
        } else {
            val readBytes = fileChannel.read(readBuffer)
            if (readBytes < 0) {
                readBuffer
            } else {
                readFile(fileChannel, readBuffer, remainingBytes - readBytes)
            }
        }
    }

    protected def withFileAtPosition(file: File, position: Long)(reader: FileChannel => ByteBuffer): ByteBuffer = {
        val fis = new FileInputStream(file)
        try {
            val channel = fis.getChannel().position(position)
            reader(channel)
        } finally {
            fis.close()
        }
    }

    // See implemented trait for description
    def rawContainerLogs(containerId: ContainerId, fromPos: Long): Future[ByteBuffer] = Future {
        blocking { // Needed due to synchronous file operations
            try {
                val file = containerLogFile(containerId)
                val size = file.length

                withFileAtPosition(file, fromPos) { channel =>
                    // Buffer allocation may fail if the log file is too large to hold in memory or
                    // too few space is left on the heap, respectively.
                    val remainingBytes = (size - fromPos).toInt
                    val readBuffer = ByteBuffer.allocate(remainingBytes)

                    readFile(channel, readBuffer, remainingBytes)
                }
            } catch {
                case e: Exception =>
                    throw new IOException(s"rawContainerLogs failed on ${containerId}", e)
            }
        }
    }
```